### PR TITLE
KAFKA-5540: Deprecate internal converter configs

### PR DESCRIPTION
**Pull request by C0urante**
Implementation of [KIP-174](https://cwiki.apache.org/confluence/display/KAFKA/KIP-174+-+Deprecate+and+remove+internal+converter+configs+in+WorkerConfig)

Configuration properties 'internal.key.converter' and 'internal.value.converter'
are deprecated, and default to org.apache.kafka.connect.json.JsonConverter.

Warnings are logged if values are specified for either, or if properties that
appear to configure instances of internal converters (i.e., ones prefixed with
either 'internal.key.converter.' or 'internal.value.converter.') are given.

The property 'schemas.enable' is also defaulted to false for internal
JsonConverter instances (both for keys and values) if it isn't specified.

Documentation and code have also been updated with deprecation notices and
annotations, respectively.

Unit tests have been updated in `PluginsTest` to account for the new defaults for `schemas.enable` for internal key/value converters, and to ensure that (for the time being), internal key/value converters are still configurable despite being deprecated.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by C0urante** [comment]
@ewencp @rhauch mind taking a look?

File: build.gradle
Comment (ewencp) [inline build.gradle]: Hmm, should this be `compileOnly` instead (and actually same for `:connect:transforms`)? We would still prefer to load these via plugin path if possible, right?

**Review by ewencp — APPROVED** [review]
Also missing something in the release and/or upgrade notes about the change.

**Comment by C0urante** [comment]
Thanks for the review, @ewencp; ready for the next round.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerConfig.java
Comment (ewencp) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerConfig.java]: There's some confusion with the versioning, but currently this is planned to be 2.0.0 due to JDK bump.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerConfig.java
Comment (ewencp) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerConfig.java]: Same change here.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerConfig.java
Comment (ewencp) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerConfig.java]: Should we log more info about the value of the setting? Specifically, I am thinking that if we see that the configured value differs from the default value, that's particularly important, and maybe even warrants logging something at a higher log level to really get the user's attention.

File: tests/kafkatest/tests/connect/templates/connect-distributed.properties
Comment (ewencp) [inline tests/kafkatest/tests/connect/templates/connect-distributed.properties:32]: Not sure if we want to maintain any tests that validate overriding this continues to work. I think in practice it's probably never worked (and never been tested) with other values; but not sure if it is valuable to make sure including the settings still work.

I suspect it's not actually worth it, but for any other similar change, I'd normally push for compatibility validation. This case is a bit unusual since schema-d converters didn't ever fully work for internal stuff afaik, but we should think through and justify whether we think tests are necessary or not.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/Plugins.java
Comment (ewencp) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/Plugins.java]: This setting doesn't seem to match the condition -- the condition checks for `internal.<x>.converter.schemas.enable` in `converterConfig`, but this is just setting `schemas.enable` in `converterConfig`. I think the condition check is the one that's wrong since `converterConfig` is generated via `originalsWithPrefix`

**Review by ewencp — COMMENTED** [review]
@C0urante A few more comments. Most are minor, only one that I think might be a correctness issue.

Sorry for the delay, but it also means this needs a bit of rebasing. But other than those issues, this is looking very close to ready to ship.

File: tests/kafkatest/tests/connect/templates/connect-distributed.properties
Comment (C0urante) [reply tests/kafkatest/tests/connect/templates/connect-distributed.properties:32]: We could beef up the `PluginsTest` unit test to ensure that overriding internal converters and/or their configs will still be picked up by the framework, but I think adding new integration/systems tests to make sure that things still work if/when that happens is unnecessary. If a different internal converter worked before these changes, and we can ensure that it will still be used in the same way by Connect after these changes, there's no reason it would stop working that would have anything to do with the changes made here. WDYT?

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerConfig.java
Comment (C0urante) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerConfig.java]: Agreed that it's worth extra emphasis if the value they've provided conflicts with the default. However--not too familiar with logging conventions in AK, so will follow your lead here, but is it really worth it to log that at the `error` level? If not, what about putting the normal deprecation messages at `info` and then upping the level to `warn` when the provided value differs from the default?

**Comment by C0urante** [comment]
@ewencp No worries about the delay :)
Made the fixes I could, asked a couple questions, ready for round 3. Will rebase after everything has been resolved.

File: build.gradle
Comment (rhauch) [inline build.gradle]: Seems strange to change the order, even if Gradle doesn't care. API is far more important, and IMO should be first, even if these are `compileOnly`.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerConfig.java
Comment (rhauch) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerConfig.java]: I do wonder if it's necessary to only log the use of a deprecated configuration property (e.g., `internal.key.converter`) when the value equals the default. Or, at least have different error messages, perhaps something like:
> INFO: Configuration `internal.key.converter` is deprecated and will be removed in an upcoming release. This value matches the current default, so it can be removed without concern.

and 
> WARN: Configuration `internal.key.converter` is deprecated and will be removed in an upcoming release. The current value `my.custom.Converter` does NOT match the default of `org.apache.kafka.connect.json.JsonConverter`, and when deprecated the behavior may change.

I'm not sure how we could log at a higher level, since technically the worker could run with the configs as they currently stand.

File: docs/upgrade.html
Comment (rhauch) [inline docs/upgrade.html]: Might be good to explain whether people need to worry about this. And I wonder if it'd be worthwhile to include the configuration properties that are now used by default so it's easier for people to see whether something has changed.

**Review by rhauch — COMMENTED** [review]
Overall, looks great. I have a few suggestions to improve the documentation and log messages.

File: docs/upgrade.html
Comment (C0urante) [reply docs/upgrade.html]: > Might be good to explain whether people need to worry about this.

Agreed, although I'm actually uncertain on the steps they should take if it turns out they do need to worry. Is it even possible to transition from one internal converter to another without wiping out the various topics (offests, config, storage) used by the Connect framework? What do we recommend if it turns out that's not actually possible?

> And I wonder if it'd be worthwhile to include the configuration properties that are now used by default so it's easier for people to see whether something has changed.

Isn't that already done with the `establishes <code>org.apache.kafka.connect.json.JsonConverter</code> with the <code>schemas.enable</code> property set to <code>false</code> as the default value for both.` part?

Comment (rhauch) [reply docs/upgrade.html]: Think of it this way: you're a user, you see this warning, and you want to know if you need to do something? You look in the documentation, and the documentation says that if your worker config contains these lines you can simply remove them now or leave them alone (and maybe get warnings in the future).

I don't think properties embedded in a paragraph are good enough. However, if you have a code block with the four lines that are equivalent to what Connect now does by default, then people will quickly recognize that's what they have and, per the documentation, conclude they can remove these lines from their worker configs without a change in behavior.

Comment (C0urante) [reply docs/upgrade.html]: I'm concerned about what we tell anyone who's using a different internal converter and/or one with `schemas.enable=true` (although I suspect that at least in the latter case just setting the value to `false` might work). If we're planning on completely removing these configurations in a future release, we should let people know now how they can deal with these changes and upgrade when the time comes.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerConfig.java
Comment (rhauch) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerConfig.java]: Okay, now that I read those messages in context, they could be improved since the subject of each sentence is different. The subject of the first sentence "Configuration {} is deprecated and will be removed in an upcoming release" is the configuration definition, whereas in the second sentence "it" refers not to the configuration but to the property set in the worker config file.

How about this:
> Worker configuration property '{}' is deprecated and may be removed in an upcoming release. The specified value '{}' matches the default, so you can safely remove this property from your worker configuration.

and
> Worker configuration property '{}' is deprecated and may be removed in an upcoming release. The specified value '{}' does NOT match the default and recommended value '{}'.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerConfig.java
Comment (rhauch) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerConfig.java]: We actually don't know what the default is for these "extra" properties.

So this logic exploded beyond what I'd hoped, and while I still think it's worthwhile to point out the case when specific properties can safely be removed, I'd be interested in whether other people find that's worth the effort.

@C0urante, is there any way of simplifying the logic? What about creating a helper method that, given the key, value, and possibly null default value, just logs the message. Then, the `logInternalConverterDeprecationWarnings` method could call this method for the various cases. Seems like it might reduce the slight repetitive logic but more importantly make it easier to see the logic that `logInternalConverterDeprecationWarnings` is using. WDYT?

File: docs/upgrade.html
Comment (rhauch) [inline docs/upgrade.html]: Need to change to 2.0

File: docs/upgrade.html
Comment (rhauch) [inline docs/upgrade.html]: How about

> In earlier releases, Connect's worker configuration required the `internal.key.converter` and `internal.value.converter` properties. In 2.0, these <a href="https://cwiki.apache.org/confluence/x/AZQ7B">are no longer required</a> and default to the JSON Converter. You may safely remove these properties from your Connect standalone and distributed worker configurations:

**Review by rhauch — COMMENTED** [review]
Thanks for the changes, @C0urante. I have a few more suggestions.

File: docs/upgrade.html
Comment (C0urante) [reply docs/upgrade.html]: Will take care of that during rebase.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerConfig.java
Comment (C0urante) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerConfig.java]: The complicated logic here is focused on catching properties prefixed with `internal.(key|value).converter`, though. A helper method for key/value pairs would simplify things but only because it wouldn't handle actually handle that.

It still seems important to catch properties that appear to be aimed at the internal converters. I'll try to simplify the logic without sacrificing that.

**Comment by C0urante** [comment]
Changed the warning semantics to only check for "extra" properties when the specified internal converter was the `JsonConverter`; otherwise, the user would already have received a warning about their converter not matching the default and any other warnings would have just been redundant.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerConfig.java
Comment (rhauch) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerConfig.java:281]: I've not compiled or tried it, but the following seems more compact, is a bit easier to see what the `logInternalConverterDeprecationWarnings(...)` method is doing, and uses consistent log messages for the different properties:

```
    private void logInternalConverterDeprecationWarnings(Map<String, String> props) {
        String[] deprecatedConfigs = new String[] {
            INTERNAL_KEY_CONVERTER_CLASS_CONFIG,
            INTERNAL_VALUE_CONVERTER_CLASS_CONFIG
        };
        for (String config : deprecatedConfigs) {
            if (props.containsKey(config)) {
                logDeprecatedProperty(config, getClass(config).getCanonicalName(), INTERNAL_CONVERTER_DEFAULT.getCanonicalName());
                // log the properties for this converter ...
                for (Map.Entry<String, Object> propEntry : originalsWithPrefix(config + ".").entrySet()) {
                    String prop = propEntry.getKey();
                    String propValue = propEntry.getValue().toString();
                    Object defaultValue = JsonConverterConfig.SCHEMAS_ENABLE_CONFIG.equals(prop) ? Boolean.FALSE : null;
                    logDeprecatedProperty(prop, propValue, defaultValue);
                }
            }
        }
    }
    private void logDeprecatedProperty(String propName, Object propValue, Object defaultValue) {
        if (defaultValue != null && defaultValue.toString().equalsIgnoreCase(propValue)) {
            log.info(
                "Worker configuration property '{}' is deprecated and may be removed in an upcoming release. "
                    + "The specified value '{}' matches the default, so this property can be safely removed from the worker configuration.",
                config,
                propValue
            );
        } else if (defaultValue != null) {
            log.warn(
                "Worker configuration property '{}' is deprecated and may be removed in an upcoming release. "
                    + "The specified value '{}' does NOT match the default and recommended value '{}'.",
                config,
                propValue,
                defaultValue
            );
        } else {
            log.warn(
                "Worker configuration property '{}' is deprecated and may be removed in an upcoming release. ",
                propKey
            );
        }
    }
```
WDYT?

**Review by rhauch — COMMENTED** [review]
One more suggestion to simplify things. Also please run the Kafka system tests and include a link to the results.

**Comment by C0urante** [comment]
@rhauch I liked the style improvements and how they made the code cleaner; however, I tweaked the logic a little bit.

If a user sees that their choice of internal converter is deprecated and doesn't match the recommended `JsonConverter`, it's unnecessary to then go on to remind them that every other property they're setting for that converter is also deprecated, especially since it's not the properties for the converter class themselves that are deprecated, but just specifying them for an internal converter.

However, if the user has specified that the internal converter should be a `JsonConverter`, it also seems helpful to let them know that _every_ property for the internal converter, as opposed to just the ones that they've included now, is deprecated; hence, inclusion of the phrase `(along with all configuration for 'internal.(key|value).converter')` in log messages.

**Comment by rhauch** [comment]
@C0urante I'd like to see some sample output for the common case where the internal converter matches the defaults. I still think the log message about a deprecated property is fine, but it seems weird to reference other properties that, IIUC, will be on adjacent log lines.

**Comment by C0urante** [comment]
@rhauch

Sample output for common case:

```
[2018-05-02 09:44:12,653] INFO Worker configuration property 'internal.key.converter' is deprecated and may be removed in an upcoming release. The specified value matches the default, so this property can be safely removed from the worker configuration. (org.apache.kafka.connect.runtime.WorkerConfig:287)
[2018-05-02 09:44:12,653] INFO Worker configuration property 'internal.key.converter.schemas.enable' (along with all configuration for 'internal.key.converter') is deprecated and may be removed in an upcoming release. The specified value matches the default, so this property can be safely removed from the worker configuration. (org.apache.kafka.connect.runtime.WorkerConfig:287)
[2018-05-02 09:44:12,653] INFO Worker configuration property 'internal.value.converter' is deprecated and may be removed in an upcoming release. The specified value matches the default, so this property can be safely removed from the worker configuration. (org.apache.kafka.connect.runtime.WorkerConfig:287)
[2018-05-02 09:44:12,653] INFO Worker configuration property 'internal.value.converter.schemas.enable' (along with all configuration for 'internal.value.converter') is deprecated and may be removed in an upcoming release. The specified value matches the default, so this property can be safely removed from the worker configuration. (org.apache.kafka.connect.runtime.WorkerConfig:287)
```

> it seems weird to reference other properties that, IIUC, will be on adjacent log lines.

The wording of the message, specifically the phrase `all configuration for`, references (or at least attempts to) properties passed on to the `internal.(key|value).converter` instance, not the class of the instance itself. It's not a redundant reminder that you shouldn't set the `internal.(key|value).converter` property, it's a notice that you should avoid setting any properties prefixed with `internal.(key|value).converter`.

**Comment by rhauch** [comment]
@C0urante I'm still not convinced the `(along with all configuration for 'internal.key.converter')` phrase is necessary or beneficial. In fact, I think it'd be simpler if it's not there. My $0.02, and would love to get other's thoughts.

**Comment by C0urante** [comment]
@ewencp care to weigh in?

**Comment by ewencp** [comment]
@C0urante @rhauch I personally would not worry about that detail (leaving it in or not). If this particular warning matters to anybody, we believe it is a very small number and if they have any questions they can ask us for more details. If people really care, the most valuable thing would be including the KIP number and/or a link to the KIP for further details.

**Comment by C0urante** [comment]
@ewencp @rhauch Rebase is complete; ready for next round of review.

File: docs/upgrade.html
Comment (rhauch) [inline docs/upgrade.html:88]: Why two bullets? Is the second sufficient?

**Comment by rhauch** [comment]
@C0urante, if you fix the HTML page comment I left above, then I'll approve as well.

**Comment by C0urante** [comment]
@rhauch sorry about that, left in accidentally. Redundant element has been removed.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/Plugins.java
Comment (ewencp) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/Plugins.java]: Is this actually safe / does it work given classloaders? Classes loaded by different classloaders are supposed to not pass `instanceof` checks against each other afaik, but I'm not sure where the `JsonConverter` comes from in this case. Basically, I'm not sure we'll actually enter this block when `JsonConverter` is loaded via `plugin.path`.

Have we tested outside of unit/integration tests that this works e2e? I think we might not hit an issue with this in practice anywhere just because connect-json ships on the same CLASSPATH as the connect-runtime, connect-api, etc, but if we ever factored out the pluggable stuff that is included w/ the framework to be loaded via separate classloaders, I think this would become a problem?

**Review by ewencp — APPROVED** [review]
This LGTM given how we ship code today, but I left one last comment that might have some impact if we adjust how we package and ship the `JsonConverter`. I'll hold off on merge until we can discuss, although I don't think it's really a blocker.

Otherwise, things LGTM, surprising that such a seemingly simple deprecation required as much careful thought and implementation as it did! Thanks for slogging through it!

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/Plugins.java
Comment (rhauch) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/Plugins.java]: Let’s test this to be sure, but a workaround for the `instanceof` test would be to check the fully qualified class names. If we go that route, it should go in a static helper method to keep the usage readable. That method would also be easier to test.

**Review by rhauch — APPROVED** [review]
LGTM, with @ewencp's `instanceof` comment notwithstanding.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/Plugins.java
Comment (C0urante) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/Plugins.java]: @ewencp @rhauch 
Originally considered nixing the `instanceof` test in favor of a comparison with the fully-qualified class name, but it came in handy within the `PluginsTest` where `JsonConverter` could be subclassed and then its configuration properties could be examined to ensure correct behavior with things like `schemas.enable` defaulting to `false`.

Also considered adding a getter method to `JsonConverter` so that the value of `enableSchemas` can be examined for testing, but adding public API that people may end up using and relying on just so that unit testing could succeed seemed like a bad tradeoff.

If there's a better way to unit test the functionality we want to establish for defaulting `schemas.enable` to false, or if it's more worth it to abandon testing in favor of using a fully qualified class name over an `instanceof` check, I'd be happy to change it. Definitely not 100% satisfied with that line as it currently stands.

**Comment by C0urante** [comment]
@ewencp @rhauch During end-to-end testing ran into a small issue involving the `Plugins` class not instantiating new converters that hadn't been explicitly configured; fixed this with the latest commit and, as long as the `instanceof JsonConverter` check is acceptable, this is ready to merge.

**Comment by rhauch** [comment]
@C0urante it might be worth manually testing the case where the JsonConverter is in a plugin, rather than on the classpath. Yeah, that's not a normal installation, but it might be good to know whether that's even a problem.

**Comment by C0urante** [comment]
@rhauch manually tested the case where the JsonConverter is available as a plugin but still on the classpath; everything worked fine.

However, if the JsonConverter is not on the classpath and only available as a plugin, the framework fails to start since the Plugins and WorkerConfig classes both depend on it at runtime.

I think requiring the JsonConverter to be on the classpath makes sense if we're going to establish it as a default converter, but if you guys disagree it'd be simple enough to remove it from the Plugins and WorkerConfig classes and hard code things instead. Going to declare the converter as a runtime dependency in the build file for now.

**Comment by rhauch** [comment]
@C0urante, @ewencp: yeah, if we're now using it by default it seems like we should expect it to be at least on the classpath. I'm good with this PR. @ewencp wanted more info on the `JsonConverter` instanceof, so he'll need to respond.

**Comment by ewencp** [comment]
retest this please

File: build.gradle
Comment (ewencp) [inline build.gradle]: Changing this to `compileOnly` breaks https://github.com/apache/kafka/blob/trunk/connect/runtime/src/main/java/org/apache/kafka/connect/tools/TransformationDoc.java#L88 which is used for autogenerating the transformation docs.

File: build.gradle
Comment (ewencp) [inline build.gradle]: This shouldn't be necessary anymore since we now have this as a `compile` dependency.

**Review by ewencp — APPROVED** [review]
I think it's fine to have it as a dependency for connect-runtime. One of the nice things about having the classpath isolation is that it not only isolates different plugins from each other, but it also isolates them from any framework/runtime dependencies as long as we haven't whitelisted them.

I think the only potentially weird behavior is that if someone subclasses `JsonConverter`, loads it from a `plugin.path` classloader, and then it hits the code path that checks `instanceof`, without whitelisting that class to always be loaded by the main classloader, the check could fail (note the unit test with `TestInternalConverter` is going to work fine since they're from the same classloader). I think we're ok here though -- that's not really the intended behavior anyway, and the only case of possible subclassing that I'm aware of anyone does in practice is the motivation for KIP-301.

I only have two nits left around the dependencies in the build script -- I'm just going to update those, push, and wait for tests to go green before merging to trunk.

**Comment by ewencp** [comment]
retest this please

**Comment by ewencp** [comment]
Hmm, not sure why Jenkins is not testing the last build, let's try this again. retest this please

**Pull request closed without merging**
